# Aula 02 (2021-01-13)
- Importação da biblioteca pandas, criação do DataFrame e eliminação das duplicatas

In [2]:
import pandas as pd
df = pd.read_csv('datasets/kc_house_data.csv')
df.drop_duplicates(subset='id', keep='last', inplace=True)

- Convertendo a coluna 'date' em datetime

In [4]:
df['date'] = pd.to_datetime(df['date'], format="%Y-%m-%d")

## Questão 1: Crie uma coluna house_age que indique se a casa é nova ou velha a partir da data de venda
- Casas vendidas a partir de 2014 serão classificadas como novas e antes disso como velhas

In [5]:
df['house_age'] = None
df.loc[df.date >= '2014-01-01', 'house_age'] = 'new_house'
df.loc[df.date < '2014-01-01', 'house_age'] = 'old_house'

- Para conferir, será mostrada aa contagem de valores para a coluna house_age

In [9]:
df['house_age'].value_counts()

new_house    21436
Name: house_age, dtype: int64

- Para o dataframe em análise, todos os imóveis são classificados como new_house, uma vez que o arquivo kc_house_data.csv possui apenas casas com datas a partir de 2014-01-01

## Questão 2: Crie uma coluna dormitory_types que classifique os imóveis a partir do número de quartos
- Imóveis com 1 quarto: Studio;
- Imóveis com 2 quartos: Apartment;
- Imóveis com mais de 2 quartos: house.

In [7]:
df['dormitory_types'] = None
df.loc[df.bedrooms == 1, 'dormitory_types'] = 'studio'
df.loc[df.bedrooms == 2, 'dormitory_types'] = 'apartment'
df.loc[df.bedrooms > 2, 'dormitory_types'] = 'house'

- Para conferir, será mostrada a contagem de valores para a coluna 'dormitory_types'

In [10]:
df['dormitory_types'].value_counts()

house        18493
apartment     2736
studio         194
Name: dormitory_types, dtype: int64

## Questão 3: Crie a coluna condition_type  para classificar os imóveis de acordo com sua condição.
- Imóveis com condição menor ou igual a 2: Bad;
- Imóveis com condição entre 3 e  4: Regular;
- Imóveis com condição 5: Good.

In [11]:
df['condition_type'] = None
df.loc[df.condition <= 2, 'condition_type'] = 'bad'
df.loc[(df.condition == 3) | (df.condition == 4), 'condition_type'] = 'regular'
df.loc[df.condition == 5, 'condition_type'] = 'good'

- Para conferir, será mostrada aa contagem de valores para a coluna contidion_type

In [14]:
df['condition_type'].value_counts()

regular    19556
good        1687
bad          193
Name: condition_type, dtype: int64

## Questão 4: Converta a coluna condition para str

In [15]:
df.condition = df['condition'].astype(str)
df.dtypes

id                          int64
date               datetime64[ns]
price                     float64
bedrooms                    int64
bathrooms                 float64
sqft_living                 int64
sqft_lot                    int64
floors                    float64
waterfront                  int64
view                        int64
condition                  object
grade                       int64
sqft_above                  int64
sqft_basement               int64
yr_built                    int64
yr_renovated                int64
zipcode                     int64
lat                       float64
long                      float64
sqft_living15               int64
sqft_lot15                  int64
house_age                  object
dormitory_types            object
condition_type             object
dtype: object

## Questão 5: Apague as colunas sqft_living15 e sqft_lot15

In [16]:
df.drop(['sqft_living15', 'sqft_lot15'], inplace=True, axis=1)

## Questão 6: Converta a coluna yr_built para datetime
- Para isso, será necessário utilizar o parâmetro format='%Y' para indicar que os valores da coluna equivalem ao ano.

In [17]:
df['yr_built'] = pd.to_datetime(df['yr_built'], format="%Y")

## Questão 7: Converta a coluna yr_renovated para datetime
- Além de ser necessário indicar pelo format que os valores da coluna equivalem ao ano, é necessário utilizar o parâmetro errors='coerce' que faz com que os valores que não consigam fazer a conversão se torne NaT (Not a Time)

In [18]:
df['yr_renovated'] = pd.to_datetime(df['yr_renovated'], format="%Y", errors="coerce")

- Conferindo os valores gerados e os tipos para as questões 6 e 7;

In [32]:
print(df[['yr_built', 'yr_renovated']].head(10))
df.dtypes

    yr_built yr_renovated
0 1955-01-01          NaT
1 1951-01-01   1991-01-01
2 1933-01-01          NaT
3 1965-01-01          NaT
4 1987-01-01          NaT
5 2001-01-01          NaT
6 1995-01-01          NaT
7 1963-01-01          NaT
8 1960-01-01          NaT
9 2003-01-01          NaT


id                          int64
date               datetime64[ns]
price                     float64
bedrooms                    int64
bathrooms                 float64
sqft_living                 int64
sqft_lot                    int64
floors                    float64
waterfront                  int64
view                        int64
condition                  object
grade                       int64
sqft_above                  int64
sqft_basement               int64
yr_built           datetime64[ns]
yr_renovated       datetime64[ns]
zipcode                     int64
lat                       float64
long                      float64
house_age                  object
dormitory_types            object
condition_type             object
dtype: object

## Questão 8: Qual a data mais antiga de construção?

In [33]:
df.yr_built.min()

Timestamp('1900-01-01 00:00:00')

**Resposta:** A data mais antiga de construção é 1900-01-01

## Questão 9: Qual a data mais antiga de renovação?

In [34]:
df.yr_renovated.min()

Timestamp('1934-01-01 00:00:00')

**Resposta:** A data mais antiga de renovação é 1934-01-01

## Questão 10: Imóveis com 2 andares.

In [35]:
df.loc[df.floors == 2, 'floors'].value_counts().item()

8209

**Resposta:** 8209 imóveis possuem 2 andares

## Questão 11: Imóveis com condição regular.

In [36]:
df.loc[df.condition_type == 'regular', 'condition_type'].value_counts().item()

19556

**Resposta:** 19556 imóveis possuem condição regular

## Questão 12: Imóveis com condição bad que possuem vista pra água

In [37]:
df.loc[(df.condition_type == 'bad') & (df.waterfront != 0), 'condition_type'].value_counts().item()

2

**Resposta:** 2 imóveis possuem vista pra água e condição bad

## Questão 13: Imóveis com condição good e new_house.

In [38]:
df.loc[(df.condition_type == 'good') & (df.house_age == 'new_house'), 'condition_type'].value_counts().item()

1687

**Resposta:** 1687 imóveis com condição good e new_house

## Questão 14: Imóvel mais caro do tipo studio

In [39]:
df.loc[(df.dormitory_types == 'studio'), 'price'].max()

1247000.0

**Resposta:** US$1247.000,00 é o preço mais caro para imóveis do tipo Studio

## Questão 15: Imóvel apartment renovado em 2015

In [40]:
df.loc[(df.dormitory_types == 'apartment') & (df.yr_renovated == '2015-01-01'), 'id'].count()

0

**Resposta:** Nenhum imóvel do tipo apartment foi renovado em 2015

## Questão 16: Qual o maior número de quartos para um imóvel do tipo house?

In [41]:
df.loc[(df.dormitory_types == 'house'), 'bedrooms'].max()

33

**Resposta:** 33 quartos

## Questão 17: Imóveis do tipo house reformados em 2014

In [43]:
df.loc[(df.house_age == 'new_house') & (df.yr_renovated == '2014-01-01'), 'id'].count()

91

**Resposta:** 91 imóveis do tipo house reformados em 2014.

## Questão 18: Selecione as colunas id, date, price, floors, zipcode de 3 formas diferentes
- Pelos índices das linhas e colunas;
- Pelos índices das linhas e nomes das colunas;
- Pelos índices das linhas e pelo tipo booleano para as colunas.

In [47]:
column = ["id", "date", "price", "floors", "zipcode"]
columnsnum = []
for i, col in enumerate(df.columns.to_list()):
    if col in column:
        columnsnum.append(i)

print(df.iloc[0:20, columnsnum])
print("-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=")
print(df.loc[0:20, column])
print("-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=")
columns = []
for i, col in enumerate(df.columns.to_list()):
    if col in column:
        columns.append(True)
    else:
        columns.append(False)
    
print(df.loc[0:20, columns])

            id       date      price  floors  zipcode
0   7129300520 2014-10-13   221900.0     1.0    98178
1   6414100192 2014-12-09   538000.0     2.0    98125
2   5631500400 2015-02-25   180000.0     1.0    98028
3   2487200875 2014-12-09   604000.0     1.0    98136
4   1954400510 2015-02-18   510000.0     1.0    98074
5   7237550310 2014-05-12  1225000.0     1.0    98053
6   1321400060 2014-06-27   257500.0     2.0    98003
7   2008000270 2015-01-15   291850.0     1.0    98198
8   2414600126 2015-04-15   229500.0     1.0    98146
9   3793500160 2015-03-12   323000.0     2.0    98038
10  1736800520 2015-04-03   662500.0     1.0    98007
11  9212900260 2014-05-27   468000.0     1.0    98115
12   114101516 2014-05-28   310000.0     1.5    98028
13  6054650070 2014-10-07   400000.0     1.0    98074
14  1175000570 2015-03-12   530000.0     1.5    98107
15  9297300055 2015-01-24   650000.0     2.0    98126
16  1875500060 2014-07-31   395000.0     2.0    98019
17  6865200140 2014-05-29   

## Questão 19: Salve um arquivo csv com as colunas de índices entre 10 e 17

In [48]:
columns = []
for i, col in enumerate(df.columns.to_list()):
    if 10 <= i <=17:
        columns.append(True)
    else:
        columns.append(False)
    
new_df = df.loc[:, columns]

new_df.to_csv('datasets/new_dataframe.csv', index=False)

## Questão 20: Salve um arquivo html com um mapa que mostre os imóveis com tamanhos distintos de acordo com o preço

In [61]:
import plotly.express as px

df_map = df[['id', 'lat', 'long', 'price', 'condition', 'date']]

mapa = px.scatter_mapbox(df_map, lat='lat', lon='long', hover_name='id', hover_data=['price'],
                        color='condition', color_continuous_scale=px.colors.sequential.Viridis, zoom=9, size='price', size_max=15)
mapa.update_layout(mapbox_style='open-street-map', margin={'l': 0, 't': 0, 'r': 0, 'b': 0})

mapa.show()

mapa.write_html('datasets/map.html')
